chat-with-rendezvous/chat.go

```diff
@@ -86,82 +91,89 @@ func main() {
 
 	// Set a function as stream handler.
-	// This function is called when a peer initiate a connection and starts a stream with this peer.
-	host.SetStreamHandler("/chat/1.1.0", handleStream)
-
-	kadDht, err := dht.New(ctx, host)
+	// This function is called when a peer initiates a connection and starts a stream with this peer.
+	host.SetStreamHandler(protocol.ID(config.ProtocolID), handleStream)
+
+	// Start a DHT, for use in peer discovery. We can't just make a new DHT client
+	// because we want each peer to maintain its own local copy of the DHT, so
+	// that the bootstrapping node of the DHT can go down without inhibitting
+	// future peer discovery.
+	kademliaDHT, err := libp2pdht.New(ctx, host)
 	if err != nil {
 		panic(err)
 	}
 
+	// Bootstrap the DHT. In the default configuration, this spawns a Background
+	// thread that will refresh the peer table every five minutes.
+	fmt.Println("Bootstrapping the DHT")
+	if err = kademliaDHT.Bootstrap(ctx); err != nil {
+		panic(err)
+	}
+
 	// Let's connect to the bootstrap nodes first. They will tell us about the other nodes in the network.
-	for _, peerAddr := range bootstrapPeers {
-		addr, _ := iaddr.ParseString(peerAddr)
-		peerinfo, _ := pstore.InfoFromP2pAddr(addr.Multiaddr())
+	for _, peerAddr := range config.BootstrapPeers {
+		peerinfo, _ := peerstore.InfoFromP2pAddr(peerAddr)
 
 		if err := host.Connect(ctx, *peerinfo); err != nil {
 			fmt.Println(err)
 		} else {
-			fmt.Println("Connection established with bootstrap node: ", *peerinfo)
+			fmt.Println("Connection established with bootstrap node:", *peerinfo)
 		}
 	}
 
 	// We use a rendezvous point "meet me here" to announce our location.
 	// This is like telling your friends to meet you at the Eiffel Tower.
-	v1b := cid.V1Builder{Codec: cid.Raw, MhType: mh.SHA2_256}
-	rendezvousPoint, _ := v1b.Sum([]byte(*rendezvousString))
-
-	fmt.Println("announcing ourselves...")
-	tctx, cancel := context.WithTimeout(ctx, time.Second*10)
-	defer cancel()
-	if err := kadDht.Provide(tctx, rendezvousPoint, true); err != nil {
-		panic(err)
-	}
+	fmt.Println("Announcing ourselves...")
+	routingDiscovery := discovery.NewRoutingDiscovery(kademliaDHT)
+	discovery.Advertise(ctx, routingDiscovery, config.RendezvousString)
+	fmt.Println("Successfully announced!")
 
 	// Now, look for others who have announced
 	// This is like your friend telling you the location to meet you.
-	// 'FindProviders' will return 'PeerInfo' of all the peers which
-	// have 'Provide' or announced themselves previously.
-	fmt.Println("searching for other peers...")
-	tctx, cancel = context.WithTimeout(ctx, time.Second*10)
-	defer cancel()
-	peers, err := kadDht.FindProviders(tctx, rendezvousPoint)
+	fmt.Println("Searching for other peers...")
+	peerChan, err := routingDiscovery.FindPeers(ctx, config.RendezvousString)
 	if err != nil {
 		panic(err)
 	}
-	fmt.Printf("Found %d peers!\n", len(peers))
 
-	for _, p := range peers {
-		fmt.Println("Peer: ", p)
-	}
+	go func() {
```

Comment:
not getting the use of this goroutine. Am I missing something or should this goroutine be inside the for-loop if you are looking to connect concurrently.